assets/js/base/components/quantity-selector/index.tsx

```diff
@@ -78,19 +81,25 @@ const QuantitySelector = ( {
 				className="wc-block-components-quantity-selector__input"
 				disabled={ disabled }
 				type="number"
-				step="1"
-				min="0"
+				step={ step }
+				min={ minimum }
 				value={ quantity }
 				onKeyDown={ quantityInputOnKeyDown }
 				onChange={ ( event ) => {
-					let value =
-						! isNumber( event.target.value ) || ! event.target.value
-							? 0
-							: parseInt( event.target.value, 10 );
+
+					let valueIsNumber = event.target.value && ( isNumber( event.target.value ) || parseInt( event.target.value, 10 ).toString() === event.target.value );
+					let value         = ! valueIsNumber ? quantity : parseInt( event.target.value, 10 );
+
 					if ( hasMaximum ) {
-						value = Math.min( value, maximum );
+						value = Math.min( value, Math.floor( maximum / step ) * step );
+					}
+
+					if ( hasStep ) {
+						value = value % step ? quantity : value;
```

Comment:
`hasStep` seems redudant and unexplantory, 1 is still a step, you should probably delete `hasStep` and move the logic here so it's `if ( step > 1 )`

Also, a comment here explaining why what's happening is preferred, since it took me a bit of time to get it.

From what I'm seeing, if the value inserted is not a multiple of `step`, we refuse it.

Reply:
Another issue I'm seeing is that we no longer can type values, if `step` is 3, you can't type in 12, or 15, because 1 is smaller than 3 and so our value would be refused.

What you should do is more `hasStep` logic to `onBlur` or debounce it somehow.

Reply:
> Another issue I'm seeing is that we no longer can type values

`quantity` is the "current" input quantity. If `step` is 3, you should be able to type 3, 6, 9, 12 because `value % step` would equal 0, and therefore `value` would preserve its value.